packages/plugin-ext/src/main/browser/command-registry-main.ts

```diff
@@ -40,7 +41,9 @@ export class CommandRegistryMainImpl implements CommandRegistryMain {
             this.delegate.registerCommand(command, {
                 // tslint:disable-next-line:no-any
                 execute: (...args: any[]) => {
-                    this.proxy.$executeCommand(command.id, ...args);
+                    // plugin command handlers cannot handle Theia URI, only VS Code URI
+                    const resolvedArgs = (args || []).map(arg => arg instanceof URI ? arg['codeUri'] : arg);
```

Comment:
looks like there is something not so neat around URI. AFAIK consumler should just use the object

Reply:
> it's not conflicting with https://github.com/theia-ide/theia/blob/master/packages/plugin-ext-vscode/src/browser/plugin-vscode-commands-contribution.ts#L46 ?

I don't think so. `vscode.open` is called by a plugin, so URI should be converted from VS Code to Theia format.

I'm trying to handle the reverse case: when Theia calls a plugin command. In this case, plugin command handlers expect only URIs in VS Code format, so Theia URIs should be always converted.

Does it make sense? Do you the better place for conversion?

Reply:
thx @akosyakov for the reply. The location of the convert is still looking strange to me here.
AFAIK If someone is calling a plug-in then this is that part that need to to convert to theia URI format, not theia plug-in checking if format that it is receiving is in a vscode format.

I'm afraid that we miss several converters. Here it is for Uri but what if there is another type, etc.
should it be inside RPC ? or is it mangled there https://github.com/theia-ide/theia/pull/4062/files

Reply:
> AFAIK If someone is calling a plug-in

The core does not know that it calls a plug-in, it calls a menu handler and pass URI of a selected resource to it. Usually it is fine. With plugin-ins it is not, because they cannot handle such URIs.

Reply:
then it should be the around menu command handler that do this kind of convert.
it should not provide something unknown for plug-ins but provide a 'good' URI

Reply:
Code in `command-registry-main` is a menu command handler from plugin-ins. Could you be more specific and suggest where it should be done if not here?

Reply:
Do you suggest to register a synthetic command for menu items which translate URIs and then delegate to plugin-in command?

Reply:
do you have an example of failing test for this usecase ?
is it only for vs code extensions or for theia plug-ins ?

Reply:
@akosyakov yes

Reply:
ok, moved to menus-contribution-handler: https://github.com/theia-ide/theia/pull/4173/files#diff-c111d7d3381b06ddff0ddb201236bec1